src/config/test.rs

```diff
@@ -0,0 +1,184 @@
+use std::{
+    fmt,
+    path::{Path, PathBuf},
+};
+
+use prometheus::proto::MetricType;
+use serde::{
+    de::{Error, MapAccess, Visitor},
+    Deserialize, Deserializer, Serialize,
+};
+
+use super::{error::TestSuiteDecodeError, Config};
+
+/// The configuration of a Quilkin testsuite.
+pub struct TestSuite {
+    pub config: Config,
+    pub options: TestConfig,
+}
+
+#[derive(Deserialize)]
+pub struct TestConfig {
+    pub config: Option<PathBuf>,
+    pub tests: std::collections::HashMap<String, TestCase>,
+}
+
+impl TestSuite {
+    pub fn find<P: AsRef<Path>>(
+        log: &slog::Logger,
+        path: Option<P>,
+    ) -> Result<Self, TestSuiteDecodeError> {
+        super::find_config_file(log, path)
+            .map_err(From::from)
+            .and_then(|s| Self::from_yaml(&s))
+    }
+
+    /// Attempts to deserialize [`Self`] from a YAML document. A valid source is
+    /// either a combination of [`Config`] document followed by [`TestConfig`]
+    /// document separated by a `---` (YAML document separator), or a
+    /// `TestConfig` document containing a `config` key that points to a valid
+    /// `Config` file.
+    pub fn from_yaml(src: &str) -> Result<Self, TestSuiteDecodeError> {
+        Ok(
+            if let Ok(options) = serde_yaml::from_str::<TestConfig>(src) {
+                let path = options
+                    .config
+                    .as_deref()
+                    .ok_or(TestSuiteDecodeError::MissingConfigInTestOptions)?;
+                let config = serde_yaml::from_reader(std::fs::File::open(path)?)?;
+                Self { config, options }
+            } else {
+                let mut de = serde_yaml::Deserializer::from_str(src);
+                let config =
+                    Config::deserialize(de.next().ok_or(TestSuiteDecodeError::MissingConfig)?)?;
+                let options = TestConfig::deserialize(
+                    de.next().ok_or(TestSuiteDecodeError::MissingTestOptions)?,
+                )?;
+                Self { config, options }
+            },
+        )
+    }
+}
+
+#[derive(Deserialize)]
+pub struct TestCase {
+    /// The data to be given to Quilkin.
+    pub input: Data,
+    /// What we expect Quilkin to send to the game server.
+    pub output: Data,
+    /// Metrics we expect to be set.
+    pub metrics: Option<std::collections::HashMap<String, MetricComparison>>,
+}
+
+#[derive(Deserialize)]
+pub struct MetricComparison {
+    pub name: String,
+    #[serde(deserialize_with = "metric_type_from_str")]
+    pub r#type: MetricType,
+    pub r#value: f64,
+}
```

Comment:
how come we're having metrics here I'm wondering how is it going to be consumed? and it seems like we're introducing our own metrics collector implelmentation, can we use the prometheus ones we export?

Reply:
This is for defining comparisons between metrics in the test file. For example if you wanted to test that the capture bytes filter captured five bytes you'd write something like the following. I have to manually deserialize to the type, because they don't provide a `FromStr` implementation for `MetricType`. It's not currently being used so I can remove it.

```yaml
tests:
   it_works:
     input: Hello
     output: Hello
     metrics:
     - name: packets_received_total
       type: gauge
       value: 5
```